### PR TITLE
picolibc: model as a feature

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Previously picolibc was selected via the `PICOLIBC=1` command line parameter.
This changes that to modelling picolibc as a platform feature (as picolibc is not available yet for all architectures supported by RIOT).

### Testing procedure

Add `FEATURES_OPTIONAL += picolibc` to your `Makefile` to opportunistically use picolibc. 

### Issues/PRs references

suggested in https://github.com/RIOT-OS/RIOT/pull/14830#discussion_r477377035

### CODE REVIEW

File: cpu/arm7_common/Makefile.dep [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # use common ARM7 periph code
 USEMODULE += arm7_common_periph
 
-ifeq (1,$(PICOLIBC))
+ifneq (,$(filter picolibc,$(FEATURES_OPTIONAL) $(FEATURES_REQUIRED)))
```
Comment (benpicco) [inline cpu/arm7_common/Makefile.dep:4]: I was afraid Travis would not like this :(

But I can't check `$(FEATURES_USED)` as in the first go-around picolibc is not in the used list yet, so newlib gets selected.

Comment (fjmolinas) [reply cpu/arm7_common/Makefile.dep:4]: Hmm that seems weird and wrong. Let me check.

Comment (fjmolinas) [reply cpu/arm7_common/Makefile.dep:4]: (wrong in the sense that the handligng is different in `info-global.inc.mk` and in `dependency_resolution.inc.mk`)

Comment (fjmolinas) [reply cpu/arm7_common/Makefile.dep:4]: See #15013.

File: kconfigs/Kconfig.features [hunk 2]
```diff
@@ -283,10 +283,10 @@ config HAS_PERIPH_WDT
     help
         Indicates that a Watchdog Timer (WDT) peripheral is present.
 
-config HAS_PERIPH_WDT_CB
+config HAS_PERIPH_PICOLIBC
```
Comment (leandrolanzieri) [inline kconfigs/Kconfig.features:286]: Why is `HAS_PERIPH_WDT_CB` removed?

Comment (benpicco) [reply kconfigs/Kconfig.features:286]: Oops, copy & paste error

**Review by fjmolinas — APPROVED** [review]
ACK, dependencies against `FEATURES_USED` and `USEMODULE` is the way to go.

**Comment by benpicco** [comment]
Hm something is missing in Riotdocker :thinking:

**Comment by benpicco** [comment]
> clang: error: argument unused during compilation: '-specs=picolibc.specs'

Looks like we have to blacklist `clang`.

**Comment by fjmolinas** [comment]
> Looks like we have to blacklist clang.

But why did this not fail before?

**Comment by benpicco** [comment]
We never compiled picolibc on CI before.

**Comment by fjmolinas** [comment]
> We never compiled picolibc on CI before.

Aha!

**Comment by aabadie** [comment]
Support was merged in riotdocker. Are all workers up-to-date ?

**Comment by aabadie** [comment]
in https://github.com/RIOT-OS/riotdocker/pull/109

**Comment by benpicco** [comment]
Yes it only fails for `TOOLCHAIN=llvm`

**Comment by fjmolinas** [comment]
I can reproduce with my local docker-image:

```
DOCKER_ENVIRONMENT_CMDLINE='-e FEATURES_OPTIONAL=picolibc -e TOOLCHAIN=llvm' BUILD_IN_DOCKER=1 BOARD=samr21-xpro make -C examples/gnrc_networking
make: Entering directory '/home/francisco/workspace/RIOT2/examples/gnrc_networking'
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/home/francisco/workspace/RIOT2:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'      -e FEATURES_OPTIONAL=picolibc -e TOOLCHAIN=llvm -e 'BOARD=samr21-xpro'  -w '/data/riotbuild/riotbase/examples/gnrc_networking/' 'riot/riotbuild:latest' make     
Building application "gnrc_networking" for "samr21-xpro" with MCU "samd21".

clang: error: argument unused during compilation: '-specs=picolibc.specs' [-Werror,-Wunused-command-line-argument]
```

**Comment by fjmolinas** [comment]
Either update and deploy the new docker image or blacklist llvm.

**Comment by fjmolinas** [comment]
> I can reproduce with my local docker-image:

I pulled just before building.

**Comment by benpicco** [comment]
Adding `CFLAGS += -Qunused-arguments` helps (or removing that from `CFLAGS`) but now we get

```
/home/benpicco/dev/RIOT/core/rmutex.c:81:17: error: address argument to atomic operation must be a pointer to a trivially-copyable type ('_Atomic(short) *' invalid)
        owner = atomic_load_explicit(&rmutex->owner, memory_order_relaxed);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/arm-none-eabi/9.2.1/include/stdatomic.h:138:5: note: expanded from macro 'atomic_load_explicit'
    __atomic_load (__atomic_load_ptr, &__atomic_load_tmp, (MO));        \
    ^              ~~~~~~~~~~~~~~~~~
/home/benpicco/dev/RIOT/core/rmutex.c:81:15: error: assigning to 'kernel_pid_t' (aka 'short') from incompatible type 'typeof (*__atomic_load_ptr)' (aka '_Atomic(short)')
        owner = atomic_load_explicit(&rmutex->owner, memory_order_relaxed);
              ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/benpicco/dev/RIOT/core/rmutex.c:107:5: error: address argument to atomic operation must be a pointer to a trivially-copyable type ('_Atomic(short) *' invalid)
    atomic_store_explicit(&rmutex->owner, thread_getpid(),
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/arm-none-eabi/9.2.1/include/stdatomic.h:126:5: note: expanded from macro 'atomic_store_explicit'
    __atomic_store (__atomic_store_ptr, &__atomic_store_tmp, (MO));     \
    ^               ~~~~~~~~~~~~~~~~~~
/home/benpicco/dev/RIOT/core/rmutex.c:146:9: error: address argument to atomic operation must be a pointer to a trivially-copyable type ('_Atomic(short) *' invalid)
        atomic_store_explicit(&rmutex->owner, KERNEL_PID_UNDEF,
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/arm-none-eabi/9.2.1/include/stdatomic.h:126:5: note: expanded from macro 'atomic_store_explicit'
    __atomic_store (__atomic_store_ptr, &__atomic_store_tmp, (MO));     \
    ^               ~~~~~~~~~~~~~~~~~~
4 errors generated.

```

**Comment by keith-packard** [comment]
Fixes to build picolibc with clang are upstream and pending for version1.4.7. You should be able to build/install current main branch for picolibc and get that working. I'm thinking that I should release that version fairly soon; we're busy getting some changes merged for CompCert (an Inria formally verified compiler) and syncing up with changes I've made in newlib's libm upstream.

**Comment by aabadie** [comment]
> syscalls.c:236:5: error: implicit declaration of function 'FDEV_SETUP_STREAM' is invalid in C99 [-Werror,-Wimplicit-function-declaration

This is fixed now: the workers have been updated on riotbuild. There are still a couple of issues on the Breeze workers, but I can't reproduce locally and it doesn't seem to be related to this PR. Maybe it also needs an update ?

File: cpu/fe310/Kconfig [hunk 3]
```diff
@@ -8,6 +8,7 @@
 config CPU_ARCH_RISCV
     bool
     select HAS_ARCH_RISCV
+    select HAS_PICOLIBC if !'$(RIOT_CI_BUILD)'
```
[the remark this answers is not in the document]
Comment (leandrolanzieri) [reply cpu/fe310/Kconfig:11]: I think comparing strings would be the way. Like so:
https://github.com/RIOT-OS/RIOT/blob/ac097bb73dc623cd4bf5deb27d5c4b82665e4f44/Kconfig#L44

Maybe `select HAS_PICOLIBC if '$(RIOT_CI_BUILD)' != '1'`?

**Comment by gschorcht** [comment]
@benpicco While developing a board definition for the [LilyGO-T-Display-GD32](https://github.com/Xinyuan-LilyGO/LilyGO-T-DisplayGD32) board, I was wondering why I got a Kconfig feature mismatch for `picolib` locally but not in CI. I guess the change in commit https://github.com/RIOT-OS/RIOT/pull/15011/commits/45270dada0af1269f8499f04e4f82329230e6944 is now obsolete.

**Pull request merged**